File: Classes/Controls/DocumentView/Text/DocumentTextViewModel.m
//
//  DocumentTextViewModel.m
//  Dokie
//
//  Created by Ashamaz Shidov on 21/12/2019.
//  Copyright © 2019 Ashamaz Shidov. All rights reserved.
//

#import "DocumentTextViewModel.h"
#import "CoreDataInclude.h"
#import "CoreDataManager.h"
#import "Logger.h"
#import "Constants.h"
#import "NavigationRouter.h"

@interface DocumentTextViewModel()

@property (nonatomic, strong) RACSubject *copyedToClipboardSubject;

@property (nonatomic, strong) NSArray<RACTuple *> *content;
@property (nonatomic, strong) NSArray<CDContent *> *displayedObjects;
@property (nonatomic, strong) CDDocument *document;

@property (nonatomic, strong, readonly) CoreDataManager *coreDataManager;
@property (nonatomic, strong, readonly) NavigationRouter *navigationRouter;

@end

@implementation DocumentTextViewModel

- (instancetype)initWithDocument:(CDDocument *)document {
    ParameterAssert(document);
    
    self = [super init];
    if (self) {
        self.document = document;
        [self setup];
    }
    return self;
}

- (void)setup {
    self.copyedToClipboardSubject = [RACSubject new];
}

- (void)addContent {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Add content", @"Add content alert title")
                                                                   message:nil
                                                            preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = NSLocalizedString(@"Description", @"Description placeholder");
        textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
    }];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = NSLocalizedString(@"Text", @"Text placeholder");
        textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
    }];
    
    @weakify(self);
    UIAlertAction *save = [UIAlertAction actionWithTitle:NSLocalizedString(@"Save", @"Save button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        NSString *loadingGuid = [self.navigationRouter showLoading];
        UITextField *detailField = alert.textFields[0];
        UITextField *textField = alert.textFields[1];
        [[[self.coreDataManager addContentToDocument:self.document withDictionary:@{kContentTypeKey : kContentTypeText,
                                                                                    kContentTextKey : textField.text ?: @"",
                                                                                    kContentDescriptionKey : detailField.text ?: @""
        }]
          concat:[self readDocument]]
         subscribeError:^(NSError * _Nullable error) {
            @strongify(self);
            [self.navigationRouter hideLoading:loadingGuid];
            [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Some error occured. Please try again later.", @"Some error title")];
        }
         completed:^{
            @strongify(self);
            [self.navigationRouter hideLoading:loadingGuid];
        }];
    }];
    [alert addAction:save];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:cancel];
    [self.navigationRouter showAlert:alert];
}

- (RACSignal *)readDocument {
    @weakify(self);
    return [[self.coreDataManager decryptedDocument:self.document]
            flattenMap:^__kindof RACSignal * _Nullable(NSArray *decrypted) {
        @strongify(self);
        NSMutableArray<RACTuple *> *content = [NSMutableArray array];
        NSMutableArray<CDContent *> *objects = [NSMutableArray array];
        for (NSInteger i = 0; i < decrypted.count; ++i) {
            RACTuple *tuple = decrypted[i];
            RACTupleUnpack(__unused CDDocument *document, NSDictionary *json, NSError *error) = tuple;
            if (!error) {
                if ([json[kContentTypeKey] isEqual:kContentTypeText]) {
                    [content addObject:RACTuplePack(json[kContentTextKey], json[kContentDescriptionKey])];
                    [objects addObject:self.document.content[i]];
                }
            }
        }
        self.content = [content copy];
        self.displayedObjects = [objects copy];
        return [RACSignal empty];
    }];
}

- (void)didSelectTextAtIndexPath:(NSIndexPath *)indexPath {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil
                                                                   message:nil
                                                            preferredStyle:UIAlertControllerStyleActionSheet];
    
    @weakify(self);
    UIAlertAction *edit = [UIAlertAction actionWithTitle:NSLocalizedString(@"Edit", @"Edit text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        [self editContentAtIndexPath:indexPath];
    }];
    [alert addAction:edit];
    UIAlertAction *copy = [UIAlertAction actionWithTitle:NSLocalizedString(@"Copy", @"Copy button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        [self copyToClipboardContentAtIndexPath:indexPath];
    }];
    [alert addAction:copy];
    UIAlertAction *delete = [UIAlertAction actionWithTitle:NSLocalizedString(@"Delete", @"Delete button text in alert view") style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        [self deleteContentAtIndexPaths:@[indexPath]];
    }];
    [alert addAction:delete];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:cancel];
    [self.navigationRouter showAlert:alert];
}

- (void)copyToClipboardContentAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.content.count) {
        RACTupleUnpack(NSString *title, NSString *detail) = self.content[indexPath.row];
        UIPasteboard *pasteboard = [UIPasteboard generalPasteboard];
        if (title.length > 0 && detail.length > 0) {
            pasteboard.string = [NSString stringWithFormat:@"%@: %@", title, detail];
        }
        else if (title.length > 0) {
            pasteboard.string = title;
        }
        else {
            pasteboard.string = detail;
        }
        [self.copyedToClipboardSubject sendNext:@YES];
    }
}

- (void)editContentAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < self.content.count) {
        RACTupleUnpack(NSString *title, NSString *detail) = self.content[indexPath.row];
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Edit content", @"Edit content alert title")
                                                                       message:nil
                                                                preferredStyle:UIAlertControllerStyleAlert];
        [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
            textField.text = title;
            textField.placeholder = NSLocalizedString(@"Description", @"Description placeholder");
            textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
        }];
        [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
            textField.text = detail;
            textField.placeholder = NSLocalizedString(@"Text", @"Text placeholder");
            textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
        }];
        
        @weakify(self);
        UIAlertAction *save = [UIAlertAction actionWithTitle:NSLocalizedString(@"Save", @"Save button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            @strongify(self);
            NSString *loadingGuid = [self.navigationRouter showLoading];
            UITextField *detailField = alert.textFields[0];
            UITextField *textField = alert.textFields[1];
            [[[self.coreDataManager updateContent:self.displayedObjects[indexPath.row]
                                   withDictionary:@{kContentTypeKey : kContentTypeText,
                                                    kContentTextKey : textField.text ?: @"",
                                                    kContentDescriptionKey : detailField.text ?: @""
                                   }]
              concat:[self readDocument]]
             subscribeError:^(NSError * _Nullable error) {
                @strongify(self);
                [self.navigationRouter hideLoading:loadingGuid];
                [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Some error occured. Please try again later.", @"Some error title")];
            }
             completed:^{
                @strongify(self);
                [self.navigationRouter hideLoading:loadingGuid];
            }];
        }];
        [alert addAction:save];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
        [alert addAction:cancel];
        [self.navigationRouter showAlert:alert];
    }
}

- (void)deleteContentAtIndexPaths:(NSArray<NSIndexPath *> *)indexPaths {
    NSMutableArray *signals = [NSMutableArray array];
    for (NSIndexPath *indexPath in indexPaths) {
        if (indexPath.row < self.displayedObjects.count) {
            [signals addObject:[self.coreDataManager deleteContent:self.displayedObjects[indexPath.row]]];
        }
    }
    if (signals.count > 0) {
        NSString *loadingGuid = [self.navigationRouter showLoading];
        @weakify(self);
        [[[RACSignal combineLatest:signals]
          concat:[self readDocument]]
         subscribeError:^(NSError * _Nullable error) {
            @strongify(self);
            [self.navigationRouter hideLoading:loadingGuid];
            [self.navigationRouter showAlertWithTitle:@"" message:NSLocalizedString(@"Some error occured. Please try again later.", @"Some error title")];
            [[self readDocument] subscribeCompleted:^{}];
        }
         completed:^{
            @strongify(self);
            [self.navigationRouter hideLoading:loadingGuid];
        }];
    }
}

#pragma mark -

- (CoreDataManager *)coreDataManager {
    return [CoreDataManager shared];
}

- (NavigationRouter *)navigationRouter {
    return [NavigationRouter shared];
}

@end

File: Classes/Views/DocumentView/Text/DocumentTextViewModel.swift
//
//  DocumentTextViewModel.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 17.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import Foundation
import ReactiveCocoa
import ReactiveSwift

class DocumentTextViewModel : NSObject {
    
    struct DocumentContent : Equatable {
        
        let text: String
        let description: String
        
        func string() -> String {
            if !text.isEmpty && !description.isEmpty {
                return "\(text): \(description)"
            }
            else if !text.isEmpty {
                return text
            }
            return description
        }
        
        static func == (lhs: Self, rhs: Self) -> Bool {
            return lhs.text == rhs.text && lhs.description == rhs.description
        }
        
    }
    
    let content = MutableProperty<[DocumentContent]>([])
    let copiedToClipboardText = MutableProperty<String?>(nil)
    var displayedObjects = [CDContent]()
    private let document: CDDocument
    private let coreDataManager = CoreDataManager.shared
    private let navigationRouter = NavigationRouter.shared
    
    @available(*, unavailable)
    override init() {
        fatalError("Not implemented")
    }
    
    init(document: CDDocument) {
        self.document = document
    }
    
    func addContent() {
        let alert = UIAlertController(title: String.localized("Add content"), message: nil, preferredStyle: .alert)
        alert.addTextField { textField in
            textField.placeholder = String.localized("Text")
            textField.autocapitalizationType = .sentences
        }
        alert.addTextField { textField in
            textField.placeholder = String.localized("Description")
            textField.autocapitalizationType = .sentences
        }
        
        let saveAction = UIAlertAction(title: String.localized("Save"), style: .default) { [unowned self] _ in
            let loadingGuid = self.navigationRouter.showLoading()
            guard let textFields = alert.textFields,
                textFields.count > 1 else { return }
            
            let textField = textFields[0]
            let detailField = textFields[1]
            self.coreDataManager
                .add(content: [C.contentTypeKey : C.contentTypeText,
                               C.contentTextKey : textField.text ?? "",
                               C.contentDescriptionKey : detailField.text ?? ""],
                     to: self.document)
                .take(during: self.reactive.lifetime)
                .observe(on: UIScheduler())
                .flatMap(.concat) { [unowned self] _ -> SignalProducer<Bool, Never> in
                    self.navigationRouter.hideLoading(loadingGuid)
                    return self.read()
            }
            .start { [unowned self] event in
                self.navigationRouter.hideLoading(loadingGuid)
                switch event {
                case .failed:
                    self.navigationRouter.showAlert(title: "", message: String.localized("Some error occured. Please try again later."))
                default:
                    break
                }
            }
        }
        let cancelAction = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
        
        alert.addAction(saveAction)
        alert.addAction(cancelAction)
        navigationRouter.showAlert(alert)
    }
    
    func didSelect(indexPath: IndexPath) {
        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        let edit = UIAlertAction(title: String.localized("Edit"), style: .default) { [unowned self] _ in
            self.edit(indexPath: indexPath)
        }
        let copy = UIAlertAction(title: String.localized("Copy"), style: .default) { [unowned self] _ in
            self.copyToClipboard(indexPath: indexPath)
        }
        let delete = UIAlertAction(title: String.localized("Delete"), style: .destructive) { [unowned self] _ in
            self.delete(indexPaths: [indexPath])
        }
        let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
        
        alert.addAction(edit)
        alert.addAction(copy)
        alert.addAction(delete)
        alert.addAction(cancel)
        
        navigationRouter.showAlert(alert)
    }
    
    func delete(indexPaths: [IndexPath]) {
        let deleteSignals = indexPaths.filter { indexPath -> Bool in
            return indexPath.row < displayedObjects.count
        }
        .map { indexPath -> SignalProducer<Bool, Error> in
            coreDataManager.delete(content: displayedObjects[indexPath.row])
        }
        
        if deleteSignals.count > 0 {
            let loadingGuid = navigationRouter.showLoading()
            SignalProducer.combineLatest(deleteSignals).take(during: reactive.lifetime)
                .observe(on: UIScheduler())
                .flatMap(.concat) { [unowned self] _ -> SignalProducer<Bool, Never> in
                    self.navigationRouter.hideLoading(loadingGuid)
                    return self.read()
            }
            .start { [unowned self] event in
                self.navigationRouter.hideLoading(loadingGuid)
                switch event {
                case .failed:
                    self.navigationRouter.showAlert(title: "", message: String.localized("Some error occured. Please try again later."))
                default:
                    break
                }
            }
        }
    }
    
    func edit(indexPath: IndexPath) {
        guard indexPath.row < content.value.count,
            indexPath.row < displayedObjects.count else { return }
        
        let value = content.value[indexPath.row]
        let alert = UIAlertController(title: String.localized("Edit content"), message: nil, preferredStyle: .alert)
        alert.addTextField { textField in
            textField.text = value.text
            textField.placeholder = String.localized("Text")
            textField.autocapitalizationType = .sentences
        }
        alert.addTextField { textField in
            textField.text = value.description
            textField.placeholder = String.localized("Description")
            textField.autocapitalizationType = .sentences
        }

        let save = UIAlertAction(title: String.localized("Save"), style: .default) { [unowned self] _ in
            guard let textFields = alert.textFields,
                textFields.count > 1 else { return }
            let loadingGuid = self.navigationRouter.showLoading()
            let textField = textFields[0]
            let detailField = textFields[1]

            self.coreDataManager.update(content: self.displayedObjects[indexPath.row], with: [C.contentTypeKey : C.contentTypeText,
                                                                                              C.contentTextKey : textField.text ?? "",
                                                                                              C.contentDescriptionKey : detailField.text ?? ""])
                .take(during: self.reactive.lifetime)
                .observe(on: UIScheduler())
                .flatMap(.concat) { [unowned self] _ -> SignalProducer<Bool, Never> in
                    self.navigationRouter.hideLoading(loadingGuid)
                    return self.read()
            }
            .start { [unowned self] event in
                self.navigationRouter.hideLoading(loadingGuid)
                switch event {
                case .failed:
                    self.navigationRouter.showAlert(title: "", message: String.localized("Some error occured. Please try again later."))
                default:
                    break
                }
            }
        }
        let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
        
        alert.addAction(save)
        alert.addAction(cancel)
        navigationRouter.showAlert(alert)
    }
    
    func read() -> SignalProducer<Bool, Never> {
        coreDataManager
            .decrypted(document: document)
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .map { [unowned self] (decryptedDocument) -> Bool in
                var content = [DocumentContent]()
                var objects = [CDContent]()
                for decrypredContent in decryptedDocument {
                    if decrypredContent.decryptionError == nil,
                        let type = decrypredContent.decrypted[C.contentTypeKey] as? String,
                        type == C.contentTypeText {
                        content.append(DocumentContent(text: decrypredContent.decrypted[C.contentTextKey] as? String ?? "",
                                                       description: decrypredContent.decrypted[C.contentDescriptionKey] as? String ?? ""))
                        objects.append(decrypredContent.content)
                    }
                }
                self.content.value = content
                self.displayedObjects = objects
                return true
        }
    }
    
    private func copyToClipboard(indexPath: IndexPath) {
        guard indexPath.row < content.value.count else { return }
        
        let value = content.value[indexPath.row]
        copiedToClipboardText.value = value.string()
        UIPasteboard.general.string = value.string()
    }
    
}
